pkg/virt-launcher/virtwrap/util/libvirt_helper.go

```diff
@@ -251,71 +254,120 @@ func StartLibvirt(stopChan chan struct{}) {
+			logfile := fmt.Sprintf("/var/log/libvirt/qemu/%s.log", domainName)
 
-			err := cmd.Start()
+			// It can take a few seconds to the log file to be created
+			for {
+				_, err = os.Stat(logfile)
+				if !os.IsNotExist(err) {
+					break
+				}
+				time.Sleep(time.Second)
+			}
+			// #nosec No risk for path injection. logfile has a static basedir
+			file, err := os.Open(logfile)
 			if err != nil {
-				log.Log.Reason(err).Error("failed to start virtlogd")
-				panic(err)
+				log.Log.Reason(err).Error("failed to catch virtlogd logs")
+				return
 			}
+			defer util.CloseIOAndCheckErr(file, nil)
 
-			go func() {
-				logfile := fmt.Sprintf("/var/log/libvirt/qemu/%s.log", domainName)
-
-				// It can take a few seconds to the log file to be created
-				for {
-					_, err = os.Stat(logfile)
-					if !os.IsNotExist(err) {
-						break
-					}
-					time.Sleep(time.Second)
-				}
-				// #nosec No risk for path injection. logfile has a static basedir
-				file, err := os.Open(logfile)
-				if err != nil {
-					log.Log.Reason(err).Error("failed to catch virtlogd logs")
-					return
-				}
-				defer util.CloseIOAndCheckErr(file, nil)
+			scanner := bufio.NewScanner(file)
+			scanner.Buffer(make([]byte, 1024), 512*1024)
+			for scanner.Scan() {
+				log.LogQemuLogLine(log.Log, scanner.Text())
+			}
 
-				scanner := bufio.NewScanner(file)
-				scanner.Buffer(make([]byte, 1024), 512*1024)
-				for scanner.Scan() {
-					log.LogQemuLogLine(log.Log, scanner.Text())
-				}
+			if err := scanner.Err(); err != nil {
+				log.Log.Reason(err).Error("failed to read virtlogd logs")
+			}
+		}()
+
+		go func() {
+			defer close(exitChan)
+			_ = cmd.Wait()
+		}()
+
+		select {
+		case <-stopChan:
+			_ = cmd.Process.Kill()
+			return
+		case <-exitChan:
+			log.Log.Errorf("virtlogd exited, restarting")
+		}
 
-				if err := scanner.Err(); err != nil {
-					log.Log.Reason(err).Error("failed to read virtlogd logs")
-				}
-			}()
+		// this sleep is to avoid consumming all resources in the
+		// event of a virtlogd crash loop.
+		time.Sleep(time.Second)
+	}
+}
 
-			go func() {
-				defer close(exitChan)
-				cmd.Wait()
-			}()
+func startQEMUSeaBiosLogging(stopChan chan struct{}) {
+	const QEMUSeaBiosDebugPipeMode uint32 = 0666
+	const logLinePrefix = "[SeaBios]:"
+
+	err := syscall.Mkfifo(QEMUSeaBiosDebugPipe, QEMUSeaBiosDebugPipeMode)
+	if err != nil {
+		log.Log.Reason(err).Error(fmt.Sprintf("%s failed creating a pipe for sea bios debug logs: %s", logLinePrefix, err.Error()))
+		return
+	}
+
+	// Chmod is needed since umask is 0018. Therefore Mkfifo does not actually create a pipe with proper permissions.
+	err = syscall.Chmod(QEMUSeaBiosDebugPipe, QEMUSeaBiosDebugPipeMode)
+	if err != nil {
+		log.Log.Reason(err).Error(fmt.Sprintf("%s failed executing chmod on pipe for sea bios debug logs: %s", logLinePrefix, err.Error()))
+		return
+	}
+
+	QEMUPipe, err := os.OpenFile(QEMUSeaBiosDebugPipe, os.O_RDONLY, 0666)
+	defer QEMUPipe.Close()
+
```

Comment:
Maybe move this to after the error handling?  Otherwise defer close() might try to close a not opened filed